### ISSUE TITLE
Problems with Chomsky Normal Form

### ISSUE THREAD

**Issue by evgenyim** [PatrikValkovic/grammpy#3]
Hello? i tried to use this library in my project, but found something interesting. According to wiki, grammar in Chomsky Normal form can't has starting nonterminal at the right side of any rule, also there can be rule S -> eps only if S is starting nonterminal. I tested your library on test S -> a S b S | eps, and found that if i creating new grammar by using transform_to_chomsky_normal_form, i found that new grammar has rules which looks like A -> S B, but i think, there shouldn't be these rules. So, i think that this algorithm doesn't work properly, or, maybe, it returns grammar in reduced CNF. Also maybe I did something wrong, so could you do the same test, please, to get sure I understood what this function should do

**Comment by PatrikValkovic** [comment]
Hi,
I will split the answer into two sections.
Firstly, the library handles epsilon rules; however, it is indeed not implemented in `ContextFree.transform_to_chomsky_normal_form` - its purpose is to split up long sequences on the right side of the rules. You need to use `ContextFree.remove_rules_with_epsilon` before [doc](https://github.com/PatrikValkovic/grammpy/blob/master/doc/transforms.md#epsilon-rules-elimination). The same applies to the unit rules.
Secondly, I checked the code, and the start symbol on the right side is not removed. It was not needed as it doesn't really matter during the CYK algorithm - it can handle it. But in any way, I understand this is not formally correct, and I will fix that as soon as possible. However, it would be rather another ContextFree transformer rather than implementing it into the `transform_to_chomsky_normal_form` itself.
Note that in any way, the grammar (or rather the transformations itself) doesn't generate formal Chomsky reduced form, because they allow *S -> eps* rule.

**Comment by evgenyim** [comment]
Thanks for your answer, now I'm starting to understand. Actually, I have one more question, ```prepare_for_cyk``` should work with any context free grammar, or not? Because i found that to use cyk we need our grammar to be in CNF, and that any grammar could be transformed to CNF, but it didn't worked in one case.

**Comment by PatrikValkovic** [comment]
Yes, the `ContextFree.prepare_for_cyk` should work on any context free grammar. What case are you talking about?

**Comment by evgenyim** [comment]
I tried on this one
S -> a S b X | a Z
X -> a Y | b Y | eps
Y -> X | c c
Z -> Z X

**Comment by PatrikValkovic** [comment]
Sorry for such a late answer @evgenyim. Regardless of the grammar, the grammar is in fact empty (you can never generate sentence out of S). Becase of the last rule Z -> Z X, the Z will always remain in the expansion. As such, the language the grammar is generating is empty.

**Comment by evgenyim** [comment]
Thanks for your answer, I understand that grammar is empty, but what if I would like to check, that it accepts something? I guess, ```prepare_for_cyk``` should return grammar anyway, but it just fails with error. I also understand, that this case won't happen often, so it's might not be necessary to fix it, but i thought that maybe it would help you to make cyk working on any context free grammar.
